### PR TITLE
SEN-2589 Add recipes to add OWASP Dependency Checker

### PR DESCRIPTION
**Pull request by seanf**
Co-authored-by: Cameron Gregor <cgregor@securecodewarrior.com>
Co-authored-by: Wander Neto <wneto@securecodewarrior.com>

### CODE REVIEW

**Review by wanderneto — APPROVED** [review]
Looks great

File: recipes/Java/Dependencies/Maven/add-build.yaml [hunk 1]
```diff
@@ -0,0 +1,42 @@
+id: scw:dependencies:maven-owasp-dep-check-add-build
+version: 10
+metadata:
+  name: Add OWASP Dependency Checker to warn about vulnerable dependencies
+  shortDescription: Add OWASP Dependency Checker to warn about vulnerable dependencies
+  level: warning
+  language: xml
+  enabled: true
+  tags: Apache Maven;security;OWASP Top 10
+search:
```
Comment (bbogason) [inline recipes/Java/Dependencies/Maven/add-build.yaml:10]: I think this should only match on pom.xml files, not every xml file that has a `<project>` tag

File: recipes/Java/Dependencies/Maven/add-build.yaml [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  actions:
+  - add:
+      childElement:
+        code: |-
+          <build>
+            <plugins>
+              <plugin>
+                <groupId>org.owasp</groupId>
+                <artifactId>dependency-check-maven</artifactId>
+                <version>6.5.0</version>
```
Comment (bbogason) [inline recipes/Java/Dependencies/Maven/add-build.yaml:29]: I am not convinced hardcoding the version here is a good idea. There seems to be new version every month.

Maybe we could consider adding a TODO here, reminding them to check for a newer version?

File: recipes/Java/Dependencies/Maven/add-build.yaml [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+                <configuration>
+                  <!-- <failBuildOnCVSS>4</failBuildOnCVSS> -->
```
Comment (bbogason) [inline recipes/Java/Dependencies/Maven/add-build.yaml:31]: Not a fan of commented out code. Consider at least adding a comment why this is commented out, and why you might want to comment this in?

**Review by bbogason — CHANGES_REQUESTED** [review]
I guess we don't have anything comparable for gradle projects?

Marking "Request changes" because `add-build` matches on every xml file (that has `<project>` tag).

The entire pom.xml is highlighted as yellow if the plugin is not configured. That feels a tad excessive? It also obscures any other warnings that might have been present in the pom file.

I tried selecting `Suppress for file` and `Suppress for tag`. That did nothing.

https://user-images.githubusercontent.com/89931719/145989213-14fc6e4e-0016-4758-acd3-0d50a92a8331.mov

**Comment by bbogason** [comment]
How widely used is this `dependency-check-maven` plugin? 
Judging by https://mvnrepository.com/artifact/org.owasp/dependency-check-maven it's not widely used at all.

**Comment by camac** [comment]
> How widely used is this `dependency-check-maven` plugin? Judging by https://mvnrepository.com/artifact/org.owasp/dependency-check-maven it's not widely used at all.

Yeah that seems very strange I find it hard to believe it is 0 all the way. I wonder if this is because it is only counting 'dependencies' in the usages? because this is used as a maven **plugin** instead of a **dependency**?
So maybe it doesn't show usages when used as a plugin?

File: recipes/Java/Dependencies/Maven/add-build.yaml (hunk 1 above)
Comment (wanderneto) [reply recipes/Java/Dependencies/Maven/add-build.yaml:10]: Fixed in this file and add-plugins.yaml

**Comment by seanf** [comment]
> > How widely used is this `dependency-check-maven` plugin? Judging by https://mvnrepository.com/artifact/org.owasp/dependency-check-maven it's not widely used at all.

> Yeah that seems very strange I find it hard to believe it is 0 all the way. I wonder if this is because it is only counting 'dependencies' in the usages? because this is used as a maven **plugin** instead of a **dependency**?
So maybe it doesn't show usages when used as a plugin?

That's exactly it. I'm not sure exactly how popular it is, but it is recommended by SCW's own challenge solutions for Java.

Here's a github search for it: https://github.com/search?l=Maven+POM&q=org.owasp+dependency-check-maven&type=Code

**Comment by seanf** [comment]
> That's exactly it. I'm not sure exactly how popular it is, but it is recommended by SCW's own challenge solutions for Java.

Oh, and I should mention that I've been using it for years. It should be used in a lot more projects.

File: recipes/Java/Dependencies/Maven/add-build.yaml (hunk 2 above)
Comment (seanf) [reply recipes/Java/Dependencies/Maven/add-build.yaml:32]: I might link to https://search.maven.org/artifact/org.owasp/dependency-check-maven

**Comment by seanf** [comment]
We'll have to target the plugins section or the build section when present (instead of project). That should cut down on the excessive highlighting, except for project which have no build section.

It's unfortunate that in XML mode we highlight the entire element - just the opening tag would be nice. When a Java class is marked, we only highlight the class name.

**Comment by seanf** [comment]
I just thought of another problem: an XML recipe can't understand the idea of parent poms. This would lead to false positives in every sub-module of a well-configured multi-module project, one where the parent pom activates the OWASP Dependency Check plugin.

Given the highlighting/suppression problems @bbogason has identified, plus the parent pom problem, I'm inclined to abandon this one. I don't think it can be done well, given Sensei's current capabilities.

**Pull request closed without merging**